[C10D] Support group_dst in scatter/gather (+object) ops

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #140847
* #140843
* __->__ #140827

Also add missing mypy typing and a few asserts to make mypy happy

Partially addresses RFC 0042 (pytorch/rfcs#71)
See more details/motivation in #140460

Note: object collective version canonicalizes to global instead of group
rank, simply becuase this left more of the original code intact and
required less conversions overall.

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @d4l3k @c-p-i-o

wconstab committed:
[C10D] Support group_dst in gather ops

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Support group_dst in gather ops"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 d4l3k c-p-i-o

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Support group_dst in gather ops"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 d4l3k c-p-i-o

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Support group_dst in gather ops"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 d4l3k c-p-i-o

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Support group_dst in gather ops"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 d4l3k c-p-i-o

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Support group_dst in gather ops"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 d4l3k c-p-i-o

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Support group_dst in gather ops"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -3826,26 +3863,22 @@ def gather(tensor, gather_list=None, dst=0, group=None, async_op=False):
-    my_rank = get_rank()
-    _validate_output_list_for_rank(my_rank, dst, gather_list)
-    output_tensors = [gather_list] if dst == my_rank else []
+    if dst is None and group_dst is None:
+        dst = 0
```
My recommendation is to error out here, instead of making assumption, in case it becomes a support burden in future.

kwen2501 replied:
oh, nvm, it is already a support burden..

kwen2501 reviewed (approved):
Overall LGTM. 
Please add PR description. It will make review easier, like why in object coll calls we choose to use global src/dst, I guess it is for avoiding double conversion and double coding?

wconstab committed:
Update on "[C10D] Support group_dst in gather ops"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 d4l3k c-p-i-o

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Support group_dst in scatter/gather (+object) ops"


Also add missing mypy typing and a few asserts to make mypy happy

Partially addresses RFC 0042 (pytorch/rfcs#71)
See more details/motivation in #140460

Note: object collective version canonicalizes to global instead of group
rank, simply becuase this left more of the original code intact and
required less conversions overall.

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Support group_dst in scatter/gather (+object) ops"


Also add missing mypy typing and a few asserts to make mypy happy

Partially addresses RFC 0042 (pytorch/rfcs#71)
See more details/motivation in #140460

Note: object collective version canonicalizes to global instead of group
rank, simply becuase this left more of the original code intact and
required less conversions overall.

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Support group_dst in scatter/gather (+object) ops"


Also add missing mypy typing and a few asserts to make mypy happy

Partially addresses RFC 0042 (pytorch/rfcs#71)
See more details/motivation in #140460

Note: object collective version canonicalizes to global instead of group
rank, simply becuase this left more of the original code intact and
required less conversions overall.

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge
